File: WordPress/Classes/Models/JetpackState.m
#import "JetpackState.h"

NSString * const JetpackVersionMinimumRequired = @"3.4.3";

@implementation JetpackState

- (NSString *)description
{
    if ([self isConnected]) {
        NSString *connectedAs = self.connectedUsername;
        if (connectedAs.length == 0) {
            connectedAs = self.connectedEmail;
        }
        if (connectedAs.length == 0) {
            connectedAs = @"UNKNOWN";
        }
        return [NSString stringWithFormat:@"🚀✅ Jetpack %@ connected as %@ with site ID %@", self.version, connectedAs, self.siteID];
    } else if ([self isInstalled]) {
        return [NSString stringWithFormat:@"🚀❌ Jetpack %@ not connected", self.version];
    } else {
        return @"🚀❔Jetpack not installed";
    }
}

- (BOOL)isInstalled
{
    return self.version != nil;
}

- (BOOL)isConnected
{
    return [self isInstalled] && [self.siteID floatValue] > 0;
}

- (BOOL)isUpdatedToRequiredVersion
{
    return [self.version compare:JetpackVersionMinimumRequired options:NSNumericSearch] != NSOrderedAscending;
}

@end

File: WordPress/Classes/Models/JetpackState.swift
@objcMembers class JetpackState: NSObject {
    static let minimumVersionRequired = "3.4.3"

    var siteID: NSNumber?
    var version: String?
    var connectedUsername: String?
    var connectedEmail: String?
    var automatedTransfer: Bool = false

    /// Returns true if Jetpack is installed and activated on the site.
    var isInstalled: Bool {
        return version != nil
    }

    /// Returns true if Jetpack is connected to WordPress.com.
    ///
    /// - Warning: Before Jetpack 3.6, a site might appear connected if it was connected and then disconnected. See https://github.com/Automattic/jetpack/issues/2137
    ///
    var isConnected: Bool {
        guard isInstalled,
            let siteID = siteID,
            siteID.intValue > 0 else {
                return false
        }
        return true
    }

    /// Returns YES if the detected version meets the app requirements.

    /// - SeeAlso: JetpackVersionMinimumRequired
    ///
    var isUpdatedToRequiredVersion: Bool {
        guard let version = version else {
            return false
        }
        return version.compare(JetpackState.minimumVersionRequired, options: .numeric) != .orderedAscending
    }

    override var description: String {
        if isConnected {
            let connectedAs = connectedUsername?.nonEmptyString()
                ?? connectedEmail?.nonEmptyString()
                ?? "UNKNOWN"
            return "🚀✅ Jetpack \(version ?? "unknown") connected as \(connectedAs) with site ID \(siteID?.description ?? "unknown")"
        } else if isInstalled {
            return "🚀❌ Jetpack \(version ?? "unknown") not connected"
        } else {
            return "🚀❔Jetpack not installed"
        }
    }
}
